src/fns.ts

```diff
@@ -219,3 +224,44 @@ export const pacemaker = async <T>(
     debug('interval cleared')
   }
 }
+
+/**
+ * Wait until the predicate returns truthy or the timeout expires.
+ * Returns a promise.
+ *
+ * Will not hang like other implementations found on NPM.
+ * Inspired by https://www.npmjs.com/package/async-wait-until
+ */
+export const waitUntil = (
+  pred: () => Promise<boolean> | boolean,
+  options: WaitOptions = {}
+) =>
+  new Promise<void>((resolve, reject) => {
+    const checkFrequencyMs = options.checkFrequency || 50
+    const timeoutMs = options.timeout || 5000
+    let checkTimer: NodeJS.Timeout
+
+    const timeoutTimer = setTimeout(() => {
+      debug('timeout')
+      clearTimeout(checkTimer)
+      reject(`Did not complete in ${timeoutMs} ms`)
+    }, timeoutMs)
+
+    const check = async () => {
+      debug('check called')
+      if (await pred()) {
+        debug('pred returned truthy')
+        clearTimeout(checkTimer)
+        clearTimeout(timeoutTimer)
+        resolve()
+      } else {
+        checkPred()
+      }
+    }
+
+    const checkPred = () => {
+      debug('checkPred called')
+      checkTimer = setTimeout(check, checkFrequencyMs)
```

Comment:
It's probably a good idea the do a `clearTimeout(checkTimer)` here before we re-assign it

Reply:
Is there a scenario that you can think of where you would have two of these timeouts running concurrently?

Reply:
Comment was just around the fact if there is a memory footprint of some sort that requires clearing a timeout even if it has fired already before re-assigning it. If that's not an issue, we don't have to add this.